engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java

```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
         }
     }
 
-    private void compileShaders(Set<ShaderProgramFeature> features) {
-        compileShader(GL20.GL_FRAGMENT_SHADER, features);
-        compileShader(GL20.GL_VERTEX_SHADER, features);
-    }
-
     private void compileAllShaderPermutations() {
-        compileShaders(Collections.<ShaderProgramFeature>emptySet());
+        int fragShaderId = compileShader(GL20.GL_FRAGMENT_SHADER, Collections.emptySet());
+        int vertShaderId = compileShader(GL20.GL_VERTEX_SHADER, Collections.emptySet());
 
-        int counter = 1;
+        if (!compileSuccess(fragShaderId) || !compileSuccess(vertShaderId)) {
 
-        for (Set<ShaderProgramFeature> permutation : Sets.powerSet(availableFeatures)) {
-            compileShaders(permutation);
-            counter++;
+            String errorMessage = String.format("Shader '%s' failed to compile.\n\n"
+                    + "Vertex Shader Info: \n%s\n"
+                    + "Fragment Shader Info: \n%s",
+                    getURI(), getLogInfo(vertShaderId), getLogInfo(fragShaderId));
+
+            throw new RuntimeException(errorMessage);
         }
-        logger.debug("Compiled {} permutations for {}.", counter, getURI());
+
+        Set<Set<ShaderProgramFeature>> allPermutations = Sets.powerSet(availableFeatures);
```

Comment:
I think here you could add the Collections.emptySet() to the allPermutions set. This would allow you to remove the whole block above, as the empty set would be handled by the loop below. You would however need to handle the empty set as a special case for the error string given to the exception.